File: tools/benchmarks/bench_pattern.ts
// Benchmark: Pattern matching on tagged union types
type Shape =
  | { kind: "circle";   radius: number }
  | { kind: "rect";     width: number; height: number }
  | { kind: "triangle"; base: number;  height: number };

function area(s: Shape): number {
  switch (s.kind) {
    case "circle":   return Math.PI * s.radius * s.radius;
    case "rect":     return s.width * s.height;
    case "triangle": return 0.5 * s.base * s.height;
  }
}

function describe(s: Shape): string {
  switch (s.kind) {
    case "circle":
      return s.radius > 10 ? "large circle" : "small circle";
    case "rect":
      return `rectangle ${s.width}x${s.height}`;
    case "triangle":
      return "triangle";
  }
}

const shapes: Shape[] = [
  { kind: "circle",   radius: 5 },
  { kind: "circle",   radius: 15 },
  { kind: "rect",     width: 4,  height: 6 },
  { kind: "rect",     width: 10, height: 3 },
  { kind: "triangle", base: 8,   height: 5 },
];

for (const s of shapes) {
  console.log(`${describe(s)} — area: ${area(s)}`);
}

File: tools/benchmarks/bench_pattern.py
# Benchmark: Pattern matching on tagged union types
from dataclasses import dataclass
from typing import Union
import math

@dataclass
class Circle:
    radius: float

@dataclass
class Rect:
    width:  float
    height: float

@dataclass
class Triangle:
    base:   float
    height: float

Shape = Union[Circle, Rect, Triangle]

def area(s: Shape) -> float:
    match s:
        case Circle(radius=r):
            return math.pi * r * r
        case Rect(width=w, height=h):
            return w * h
        case Triangle(base=b, height=h):
            return 0.5 * b * h

def describe(s: Shape) -> str:
    match s:
        case Circle(radius=r) if r > 10:
            return "large circle"
        case Circle():
            return "small circle"
        case Rect(width=w, height=h):
            return f"rectangle {w}x{h}"
        case Triangle():
            return "triangle"

shapes: list[Shape] = [
    Circle(5.0), Circle(15.0),
    Rect(4.0, 6.0), Rect(10.0, 3.0),
    Triangle(8.0, 5.0),
]

for s in shapes:
    print(f"{describe(s)} — area: {area(s)}")
